### ISSUE TITLE
SQL code generation and redundant calls in cycle for child entity list

### ISSUE THREAD

**Issue by npochechuev** [linq2db/linq2db#288]
I didn't find good solution to prevent redundant DB calls without manual mapping when my entity has child list of entities 

```cs
                var test1 = (from ch in db.ParentEntity
                            select new
                            {
                                ID = ch.ID,
                                ...
                                TestChildHisory = db.ChildEntity.Where(lb => ch.ID == lb.ParentEntityID ).Select(z => new
                                {
                                    ID = z.ID,
                                    ParentEntityID = z.ParentEntityID ,
                                    SomeAttr = z.SomeAttr,

                                }).ToList()

                            }
                    );

```

I can use the next query with manual mapping only.

```cs

                var test = (
                    from ch in db.ParentEntity
                    join ch_fts in db.ChildEntity
                        on ch.ID equals ch_fts.ParentEntityID
                        into outer1
                    from item1 in outer1.DefaultIfEmpty()
                    orderby ch.ID

                    select new
                    {
                        ID = ch.ID,
                        SomeParentAttr = ch.SomeParentAttr,
                        SomeParentAttr2 = ch.SomeParentAttr2,
                        ID_child = item1.ID,

                        SomeChildAttr = item1.SomeChildAttr,
                        SomeChildAttr2 = item1.SomeChildAttr2
                    });

                Int64 currentParentID = -1;
                var outputList = new List<ParentEntity> ();
                ParentEntity currentTestParentEntity = new ParentEntity();

                foreach (var chH in test)
                {
                    if (currentParentID != chH.ID)
                    {
                        currentTestParentEntity = new TestCharacter() {ID = chH.ID, ..., ChildEntityList = new List<TestChildEntity>()};
                        outputList.Add(currentTestParentEntity);
                        currentParentID = currentTestParentEntity.ID
                    }
                    currentTestParent.ChildEntityList.Add(new TestChildEntity{ID = chH.ID, ...});
                }
```

Usually EntityFramework generates single denormalized query in this case. When parent entity contains few child lists it generates select-query with "union all" and returns full result in single dataset for example 

```sql
SELECT 
    [UnionAll1].[CurrentXP] AS [C1], 
    [UnionAll1].[ID] AS [C2], 
    [UnionAll1].[CurrentXP1] AS [C3], 
    [UnionAll1].[C1] AS [C4], 
    [UnionAll1].[C2] AS [C5], 
    [UnionAll1].[ID1] AS [C6], 
    [UnionAll1].[CharacterID] AS [C7], 
    [UnionAll1].[DateOfCheckIn] AS [C8], 
    [UnionAll1].[C3] AS [C9], 
    [UnionAll1].[C4] AS [C10], 
    [UnionAll1].[C5] AS [C11], 
    [UnionAll1].[C6] AS [C12]
    FROM  (SELECT 
        CASE WHEN ([Extent2].[ID] IS NULL) THEN CAST(NULL AS int) ELSE 1 END AS [C1], 
        [Extent1].[CurrentXP] AS [CurrentXP], 
        [Extent1].[ID] AS [ID], 
        [Extent1].[CurrentXP] AS [CurrentXP1], 
        CASE WHEN ([Extent2].[ID] IS NULL) THEN CAST(NULL AS int) ELSE 1 END AS [C2], 
        [Extent2].[ID] AS [ID1], 
        [Extent2].[CharacterID] AS [CharacterID], 
        [Extent2].[DateOfCheckIn] AS [DateOfCheckIn], 
        CAST(NULL AS int) AS [C3], 
        CAST(NULL AS bigint) AS [C4], 
        CAST(NULL AS bigint) AS [C5], 
        CAST(NULL AS bigint) AS [C6]
        FROM  [dbo].[Character] AS [Extent1]
        LEFT OUTER JOIN [dbo].[CheckInHistory] AS [Extent2] ON [Extent1].[ID] = [Extent2].[CharacterID]
    UNION ALL
        SELECT 
        2 AS [C1], 
        [Extent3].[CurrentXP] AS [CurrentXP], 
        [Extent3].[ID] AS [ID], 
        [Extent3].[CurrentXP] AS [CurrentXP1], 
        CAST(NULL AS int) AS [C2], 
        CAST(NULL AS bigint) AS [C3], 
        CAST(NULL AS bigint) AS [C4], 
        CAST(NULL AS datetime2) AS [C5], 
        2 AS [C6], 
        [Extent4].[ID] AS [ID1], 
        [Extent4].[CharacterID] AS [CharacterID], 
        [Extent4].[UpgradeID] AS [UpgradeID]
        FROM  [dbo].[Character] AS [Extent3]
        INNER JOIN [dbo].[CharacterUpgrade] AS [Extent4] ON [Extent3].[ID] = [Extent4].[CharacterID]) AS [UnionAll1]
    ORDER BY [UnionAll1].[ID] ASC, [UnionAll1].[C1] ASC
```

**Comment by MaceWindu** [comment]
Yeah, it would be nice to have it automated.
I do such loads with 1+N(number of child collections) queries using following approach.

```
var parentsQuery = db.Parents.Where(some => filter);

var parents = parentsQuery.OrderBy(desired => sort).ToList();

var childsDictionary = db.Childs
    .Where(c => parentsQuery.Select(p => p.Id).Contains(c.parentId))
    .ToDictionary(c => c.parentId, c => new ChildDTO() {...});

foreach (parent in parents)
{
    parent.Childs = ChildsDictionary[parent.Id];
}
```
Comparing to your approach it should be less performant due to extra query, with less code on client.

**Comment by npochechuev** [comment]
@MaceWindu,

It looks like you use the same logic with cycles like use LINQ query in my 1st example.

@igor-tkachev,
BTW,
May be the next logic would be better when we need denormalized result in single dataset.

```
SELECT  [Extent1].[CurrentXP] AS [CurrentXP], 
        [Extent1].[ID] AS [ID], 
        [Extent1].[CurrentXP] AS [CurrentXP1],
		ds.*,

		[Extent2].[ID] AS [C2], 
        [Extent2].[ID] AS [ID1], 
        [Extent2].[CharacterID] AS [CharacterID], 
        [Extent2].[DateOfCheckIn] AS [DateOfCheckIn],

		[Extent4].[ID] AS [ID1], 
        [Extent4].[CharacterID] AS [CharacterID], 
        [Extent4].[UpgradeID] AS [UpgradeID]

  FROM  [dbo].[Character] AS [Extent1]
  --finish main dataset and creating datasets for child list
 cross
  join (select 1 as datasetNumber union all select 2 union all select 3) ds

  --Start dataset 2
  left
  join [dbo].[CheckInHistory] AS [Extent2]
    on ds.datasetNumber = 2 
   and [Extent1].[ID] = [Extent2].[CharacterID]
  
  --Start dataset 3
  left
  join [dbo].[CharacterUpgrade] AS [Extent4] 
    on ds.datasetNumber = 3
   and [Extent1].[ID] = [Extent4].[CharacterID]

where ds.datasetNumber = 1
   or (ds.datasetNumber = 2 and  [Extent2].[CharacterID] is not null)
   or (ds.datasetNumber = 3 and  [Extent4].[CharacterID] is not null)

order by [Extent1].[ID]
```

**Comment by npochechuev** [comment]
if you need the same amount of row as for EF and the same logic you can use the next logic 

```
SELECT  [Extent1].[CurrentXP] AS [CurrentXP], 
        [Extent1].[ID] AS [ID], 
        [Extent1].[CurrentXP] AS [CurrentXP1],
        ds.*,

        [Extent2].[ID] AS [C2], 
        [Extent2].[ID] AS [ID1], 
        [Extent2].[CharacterID] AS [CharacterID], 
        [Extent2].[DateOfCheckIn] AS [DateOfCheckIn],

        [Extent4].[ID] AS [ID1], 
        [Extent4].[CharacterID] AS [CharacterID], 
        [Extent4].[UpgradeID] AS [UpgradeID]

  FROM  [dbo].[Character] AS [Extent1]
  --finish main dataset and creating datasets for child list
 cross
  join (select 2 as datasetNumber union all select 3) ds

  --Start dataset 2
  left
  join [dbo].[CheckInHistory] AS [Extent2]
    on ds.datasetNumber = 2 
   and [Extent1].[ID] = [Extent2].[CharacterID]

  --Start dataset 3
  left
  join [dbo].[CharacterUpgrade] AS [Extent4] 
    on ds.datasetNumber = 3
   and [Extent1].[ID] = [Extent4].[CharacterID]

where (
       ds.datasetNumber = 2
       or (ds.datasetNumber = 3 and  [Extent4].[CharacterID] is not null)
	  )
order by [Extent1].[ID]
```

**Comment by ili** [comment]
I think it would be a part of v2

**Comment by MaceWindu** [comment]
Implemented by #1756
